File: projects/embervm/control/test/embervm/s3_client/sigv4_test.exs
defmodule Embervm.S3Client.SigV4Test do
  use ExUnit.Case, async: true

  alias Embervm.S3Client.SigV4

  test "AWS documented GET object known answer" do
    headers = [
      {"range", "bytes=0-9"},
      {"x-amz-content-sha256", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"},
      {"x-amz-date", "20130524T000000Z"}
    ]

    signed_headers = ["host", "range", "x-amz-content-sha256", "x-amz-date"]
    body_sha = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"

    canonical =
      SigV4.canonical_request(
        :get,
        "https://examplebucket.s3.amazonaws.com/test.txt",
        headers,
        signed_headers,
        body_sha
      )

    scope = "20130524/us-east-1/s3/aws4_request"
    canonical_hash = :crypto.hash(:sha256, canonical) |> Base.encode16(case: :lower)
    string_to_sign = "AWS4-HMAC-SHA256\n20130524T000000Z\n#{scope}\n#{canonical_hash}"

    assert SigV4.signature(
             "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
             "20130524",
             "us-east-1",
             "s3",
             string_to_sign
           ) == "f0e8bdb87c964420e857bd35b5d6ed310bd44f0170aba48dd91039c6036bdb41"
  end

  test "credentials add SigV4 headers and empty credentials are unchanged" do
    now = ~U[2026-08-21 12:34:56Z]
    url = "http://seaweedfs/embervm/?prefix=base%2Famd%2F&list-type=2"
    creds = %{access_key_id: "embervm", secret_access_key: "secret"}
    headers = SigV4.sign(:get, url, [], SigV4.unsigned_payload(), creds, now)

    authorization = headers |> Map.new() |> Map.fetch!("authorization")
    assert authorization =~ ~r/^AWS4-HMAC-SHA256 Credential=embervm\/20260821\/us-east-1\/s3\/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=[0-9a-f]{64}$/
    assert Map.new(headers)["x-amz-content-sha256"] == "UNSIGNED-PAYLOAD"
    assert SigV4.sign(:get, url, [], SigV4.unsigned_payload(), %{}, now) == []
  end

  test "canonical query encodes slashes in keys and values the way S3 recomputes them" do
    # SeaweedFS canonicalises the raw query with the AWS unreserved set, so a raw
    # `delimiter=/` must sign as `delimiter=%2F` or the signature never matches.
    canonical = SigV4.canonical_request(:get, "http://seaweedfs/embervm/?prefix=a/b&delimiter=/", [], ["host"], "UNSIGNED-PAYLOAD")
    [_method, _uri, query | _] = String.split(canonical, "\n")
    assert query == "delimiter=%2F&prefix=a%2Fb"
  end
end

File: projects/embervm/noded/store/sigv4_test.go
package store

import (
	"context"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestSigV4AWSKnownAnswer(t *testing.T) {
	req, err := http.NewRequest(http.MethodGet, "https://examplebucket.s3.amazonaws.com/test.txt", nil)
	if err != nil {
		t.Fatal(err)
	}
	req.Header.Set("Range", "bytes=0-9")
	req.Header.Set("x-amz-content-sha256", "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")
	req.Header.Set("x-amz-date", "20130524T000000Z")
	signedHeaders := []string{"host", "range", "x-amz-content-sha256", "x-amz-date"}
	canonical := canonicalRequest(req, signedHeaders, req.Header.Get("x-amz-content-sha256"))
	scope := "20130524/us-east-1/s3/aws4_request"
	stringToSign := "AWS4-HMAC-SHA256\n20130524T000000Z\n" + scope + "\n" + sha256Hex(canonical)
	got := calculateSignature(
		"wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY",
		"20130524", "us-east-1", "s3", stringToSign,
	)
	const want = "f0e8bdb87c964420e857bd35b5d6ed310bd44f0170aba48dd91039c6036bdb41"
	if got != want {
		t.Fatalf("signature = %s, want %s\ncanonical request:\n%s", got, want, canonical)
	}
}

func TestSignNoCredentialsLeavesRequestUntouched(t *testing.T) {
	s := New("http://example.test", "embervm", false)
	req, err := http.NewRequest(http.MethodGet, s.url("base/a/meta.json"), nil)
	if err != nil {
		t.Fatal(err)
	}
	before := req.Header.Clone()
	if err := s.sign(req); err != nil {
		t.Fatal(err)
	}
	if len(req.Header) != len(before) || req.Header.Get("Authorization") != "" || req.Header.Get("x-amz-date") != "" {
		t.Fatalf("anonymous signing changed headers: %#v", req.Header)
	}
}

func TestAnonymousRequestSendsNoAuthorizationHeader(t *testing.T) {
	authorization := make(chan string, 1)
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		authorization <- req.Header.Get("Authorization")
		w.WriteHeader(http.StatusOK)
	}))
	defer srv.Close()
	s := New(srv.URL, "embervm", false)
	if ok, err := s.Head(context.Background(), "base/amd/demo/meta.json"); err != nil || !ok {
		t.Fatalf("Head = %v, %v", ok, err)
	}
	if got := <-authorization; got != "" {
		t.Fatalf("Authorization = %q, want absent", got)
	}
}

func TestSignUsesPinnedClock(t *testing.T) {
	s := New("http://example.test", "embervm", false, WithCredentials("id", "secret"))
	s.now = func() time.Time { return time.Date(2026, 8, 21, 12, 34, 56, 0, time.UTC) }
	req, _ := http.NewRequest(http.MethodGet, s.url("base/a/meta.json"), nil)
	if err := s.sign(req); err != nil {
		t.Fatal(err)
	}
	if got := req.Header.Get("x-amz-date"); got != "20260821T123456Z" {
		t.Fatalf("x-amz-date = %q", got)
	}
}
